In [1]:
from sklearn.datasets import load_iris

In [2]:
iris = load_iris()
type(iris)

#bunch is scikit learn special dataset that stores its attributes


sklearn.datasets.base.Bunch

In [3]:
print(iris.data)

[[ 5.1  3.5  1.4  0.2]
 [ 4.9  3.   1.4  0.2]
 [ 4.7  3.2  1.3  0.2]
 [ 4.6  3.1  1.5  0.2]
 [ 5.   3.6  1.4  0.2]
 [ 5.4  3.9  1.7  0.4]
 [ 4.6  3.4  1.4  0.3]
 [ 5.   3.4  1.5  0.2]
 [ 4.4  2.9  1.4  0.2]
 [ 4.9  3.1  1.5  0.1]
 [ 5.4  3.7  1.5  0.2]
 [ 4.8  3.4  1.6  0.2]
 [ 4.8  3.   1.4  0.1]
 [ 4.3  3.   1.1  0.1]
 [ 5.8  4.   1.2  0.2]
 [ 5.7  4.4  1.5  0.4]
 [ 5.4  3.9  1.3  0.4]
 [ 5.1  3.5  1.4  0.3]
 [ 5.7  3.8  1.7  0.3]
 [ 5.1  3.8  1.5  0.3]
 [ 5.4  3.4  1.7  0.2]
 [ 5.1  3.7  1.5  0.4]
 [ 4.6  3.6  1.   0.2]
 [ 5.1  3.3  1.7  0.5]
 [ 4.8  3.4  1.9  0.2]
 [ 5.   3.   1.6  0.2]
 [ 5.   3.4  1.6  0.4]
 [ 5.2  3.5  1.5  0.2]
 [ 5.2  3.4  1.4  0.2]
 [ 4.7  3.2  1.6  0.2]
 [ 4.8  3.1  1.6  0.2]
 [ 5.4  3.4  1.5  0.4]
 [ 5.2  4.1  1.5  0.1]
 [ 5.5  4.2  1.4  0.2]
 [ 4.9  3.1  1.5  0.1]
 [ 5.   3.2  1.2  0.2]
 [ 5.5  3.5  1.3  0.2]
 [ 4.9  3.1  1.5  0.1]
 [ 4.4  3.   1.3  0.2]
 [ 5.1  3.4  1.5  0.2]
 [ 5.   3.5  1.3  0.3]
 [ 4.5  2.3  1.3  0.3]
 [ 4.4  3.2  1.3  0.2]
 [ 5.   3.5

# Machine Learning Terminology


Each row is an observation (also known as: sample, example, instance, record)

Each column is a feature (also known as: preditor, attribute, independent variable, input, regressor, covariate)

In [4]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [5]:
print(iris.target)

#this prints integers representing the integers of the species of each observation
#Target represents what we are going to predict
# 0 , represents setosa
#1, versicolor
#2, virginica

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [6]:
print(iris.target_names)

['setosa' 'versicolor' 'virginica']


Each value we are predicting is the response (also known as: target, outcome, label, dependent varaible)

Classification is supervised learning in which the response is ordered and continuous 

Regression is supervised learning in which the response is ordered and continuous

# Requirements for working with data in scikit-learn

Features and response are separate objects

Features and response should be numeric (output should be numeric, ex. 0,1,2)

Features and response should be NumPy arrays

Features and response should have specific shapes

In [7]:
# check out the types of features and response

print(type(iris.data))
print(type(iris.target))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [8]:

# check the shape of the features (first dimension = number of observations, 
#second dimensions = number of features)
# check the shape of the response (single dimension matching the number of observations)
print(iris.target.shape)
print(iris.data.shape)

(150,)
(150, 4)


In [13]:

# check the shape of the response (single dimension matching the number of observations)
print(iris.target.shape)

(150,)


In [14]:
# store feature matrix in "X", X is upper case bc it represents a matrix
X = iris.data

# store response vector in "y", y is lowercase because it represents a vector
y = iris.target

150 observations

4 features (sepal length, sepal width, petal length, petal width)

Response variable is the iris species

Classification problem since response is categorical

In [15]:

# print the shapes of X and y
print(X.shape)
print(y.shape)

(150, 4)
(150,)


# K-nearest neighbors (KNN) classification



1. Pick a value for K.

2. Search for the K observations in the training data that are "nearest" to the measurements of the unknown iris.

3. Use the most popular response value from the K nearest neighbors as the predicted response value for the unknown iris.

# Scikit - learn 4 step modeling pattern


Step 1: Import the class you plan to use

In [18]:

from sklearn.neighbors import KNeighborsClassifier

Step 2: "Instantiate" the "estimator"

"Estimator" is scikit-learn's term for model

"Instantiate" means "make an instance of"

In [20]:
knn = KNeighborsClassifier(n_neighbors=1)

#we have instantiated the classifier
#n_neighbors = 1 means looks for the closest neighbor

Name of the object does not matter

Can specify tuning parameters (aka "hyperparameters") during this step

All parameters not specified are set to their defaults

In [21]:

print(knn)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')



Step 3: Fit the model with data (aka "model training")

Model is learning the relationship between X and y

Occurs in-place

In [22]:
knn.fit(X, y)

#use fit method, pass the matrix and vector

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

Step 4: Predict the response for a new observation

New observations are called "out-of-sample" data

Uses the information it learned during the model training process

In [23]:

knn.predict([[3, 5, 4, 2]])

#asking knn model to predict the series, then pass it to a list

array([2])

Returns a NumPy array

Can predict for multiple observations at once

In [24]:
#x_new contains 2 new observations, with 4 features each
X_new = [[3, 5, 4, 2], [5, 4, 3, 2]]
knn.predict(X_new)

array([2, 1])

# Using different Value for K

In [26]:

# instantiate the model (using the value K=5)
knn = KNeighborsClassifier(n_neighbors=5)

# fit the model with data
knn.fit(X, y)

# predict the response for new observations
knn.predict(X_new)

array([1, 1])

# Using a different classification model

In [27]:
# import the class
from sklearn.linear_model import LogisticRegression

# instantiate the model (using the default parameters)
logreg = LogisticRegression()

# fit the model with data
logreg.fit(X, y)

# predict the response for new observations
logreg.predict(X_new)

array([2, 0])